File: problems/0004_median-of-two-sorted-arrays/binary-search/Solution.php
<?php
class Solution {
    /**
     * @param Integer[] $nums1
     * @param Integer[] $nums2
     * @return Float
     */
    function findMedianSortedArrays($nums1, $nums2) {
        $A = $nums1;
        $B = $nums2;
        $m = count($A);
        $n = count($B);
        if ($m > $n) {
            $A = $nums2;
            $B = $nums1;
            $tmp = $m;
            $m = $n;
            $n = $tmp;
        }
        $total = $m + $n;
        $h = intdiv($total, 2);
        $isOdd = ($total % 2 == 1);
        for ($iBegin = 0, $iEnd = $m; $iBegin <= $iEnd;) {
            $i = intdiv($iBegin + $iEnd, 2);
            $j = $h - $i;
            if ($i < $iEnd && 0 < $j && $A[$i] < $B[$j - 1])
                $iBegin = $i + 1;
            else if ($iBegin < $i && $j < $n && $B[$j] < $A[$i - 1])
                $iEnd = $i - 1;
            else {
                $minGreater = ($i == $m ?
                        $B[$j] :
                        ($j == $n ?
                            $A[$i] :
                            min($A[$i], $B[$j])));
                if ($isOdd)
                    return $minGreater;
                $maxLesser = ($i == 0 ?
                        $B[$j - 1] :
                        ($j == 0 ?
                            $A[$i - 1] :
                            max($A[$i - 1], $B[$j - 1])));
                return ($maxLesser + $minGreater) / 2;
            }
        }
        throw new Exception("Should NOT be HERE!");
    }
}

File: problems/0004_median-of-two-sorted-arrays/binary-search/Solution.swift
class Solution {
    func findMedianSortedArrays(_ nums1: [Int], _ nums2: [Int]) -> Double {
        let (A, B) = (nums1.count <= nums2.count ?
                (nums1, nums2) : (nums2, nums1))
        let (m, n) = (A.count, B.count)
        let total = m + n
        let (h, isOdd) = (total / 2, total % 2 == 1)
        var (iBegin, iEnd) = (0, m)
        while iBegin <= iEnd {
            let i = (iBegin + iEnd) / 2
            let j = h - i
            if i < iEnd && 0 < j && A[i] < B[j - 1] {
                iBegin = i + 1
            } else if iBegin < i && j < n && B[j] < A[i - 1] {
                iEnd = i - 1
            } else {
                let minGreater = (i == m ?
                        B[j] :
                        (j == n ?
                            A[i] :
                            min(A[i], B[j])))
                if isOdd {
                    return Double(minGreater)
                }
                let maxLesser = (i == 0 ?
                        B[j - 1] :
                        (j == 0 ?
                            A[i - 1] :
                            max(A[i - 1], B[j - 1])))
                return Double(maxLesser + minGreater) / 2.0
            }
        }
        fatalError("Should NOT be HERE!")
    }
}
